[cmake/ios] Enable Xcode generator for IOS

fetzerch opened the pull request:
Not yet tested, as there's still something wrong with my local ios depends.
cmake run succeeds at least. 

@Memphiz I'll let you know once I could verify the build.

Memphiz commented:
@fetzerch it creates the xcode project now. Next error is this:

http://pastebin.com/uYk903jQ

It somehow tries to compile an universal binary (which is also called fat binary and can be used to have multiple architectures into one dylib or executable at the same time). It fails because it tries to add the armv7 architecture into the universal binary 2 times (only one architecture can be in a fat binary once). Basically we don't want to and can't use universal binaries at all (because depends only builds for one single architecture).
Thats really strange. Wasn't cmake supposed to get this stuff right? I mean why does it work ok with makefiles but becoming such a bitch for xcode (which should only be a different generator and not change behavior at all)? ;)

akva2 commented:
because Makefiles are sane and xcode is a huge pile of hack. /runs

Memphiz commented:
@akva2 yep - maybe - but cmake is supposed to abstract the huge pile of hack ;)

akva2 commented:
touchè

MrMC commented:
HAHAHAHAHAHA

MrMC commented:
Just an FYI, you can built a universal binary. Depends supports it, Xcode supports it, just have to package it right. We did an experimental built for MrMC on iOS to support armv7 and arm64 under one binary using embedded bitcode. Was eventually abandoned as hitting armv7 devices were just not worth it.

fetzerch committed:
[cmake] Set correct output directory for core_add_shared_library

Set the correct output-directory for shared libraries on all platforms
(module vs. shared library with different extensions on Darwin) and
for all configurations (Debug, Release, RelWithDebInfo, MinSizeRelease).

fetzerch committed:
[cmake/ios] Enable Xcode generator for IOS

- Adapt Toolchain file to build a bundles with xcode.
  This change has to be done in the Toolchain file because CMake
  otherwise fails to run some compile checks.
- Unify path where bundle is built for Makefiles and Xcode
  to build/$(CONFIGURATION)-$(EFFECTIVE_PLATFORM_NAME)
- Set only one architecture to prevent building multi-arch bundles.
- Simplify Install.cmake. Existing scrips are called now as post build
  steps so that the bundle can be executed with Xcode
- Apply xcconfig settings.

Note: Two adaptions to the existing "autotools/Xcode" scripts were
needed.
- CMake cannot have the bundle name be different to the binary
  and we cannot call the binary "Kodi.bin". The script has been adapted
  to adapt the library execution path directly in place.
- CMake calls mkdeb.sh with "Debug-iphoneos", hence the script has been
  adapted to match strings starting with "debug" or "release".

Reference:
 - https://cmake.org/Bug/view.php?id=15329
 - http://stackoverflow.com/questions/33660608/how-can-i-disable-xcode-bitcode-within-a-cmake-project

fetzerch commented:
@Memphiz: Got it working now. Main issue was that for ios, xcode can only build bundles. This required to change the way I used to control the packaging in Install.cmake. Then there were a few minor issues (see commit message). I've successfully built and run the app on the ipad from xcode. In addition building the deb works now with make and xcodebuild and after signing I can install and run both.

Could you give it another try? After opening the xcode project, you still need to select the signing properties (bundle identifier, team). And since the toolchain file is touched, it needs a reconfigure of depends.

jenkins build this please with addons

Memphiz commented:
Nice - will give it a shot asap. Jenkins seems unhappy? Or just a fluke?

fetzerch commented:
Right, the changes to the toolchain must be only done when building with xcode.
The other errors seem unrelated, but let's try again.

jenkins build this please with addons

fetzerch commented:
Jenkins build is fine now as well (http://jenkins.kodi.tv/job/IOS/8546)

Memphiz commented:
Great. Will give it a try the next days and adapt the osx/ios readmes to reflect the new workflow

Memphiz commented on project/cmake/scripts/ios/ArchSetup.cmake:
```diff
@@ -13,6 +13,7 @@ if(WITH_ARCH)
   set(ARCH ${WITH_ARCH})
 else()
   if(CPU STREQUAL armv7 OR CPU STREQUAL arm64)
+    set(CMAKE_OSX_ARCHITECTURES armv7)
```
set CMAKE_OSX_ARCHITECTURES CPU? - we want to be able to build arm64 aswell ... also where does the CPU come from? i thought cmake variables must be references by ${CPU} then?

Memphiz commented:
nevermind my comments - hadn't reset my branch to you remote ... works fine for me on xcode7.3 so far.

fetzerch commented:
arm64 build works as well now.
jenkins build this please with addons

Memphiz commented:
Nice :)

fetzerch commented:
come on jenkins, you can do it ;)
jenkins build this with addons please

fetzerch commented:
jenkins build this with addons please

fetzerch commented:
jenkins build this with addons please

fetzerch commented:
jenkins build this with addons please
